File: direct/1C.py
from math import*
p=[list(map(float,input().split()))for i in range(3)]
a,b,c=[hypot(x1-x2,y1-y2)for(x1,y1),(x2,y2)in[(p[i],p[(i+1)%3])for i in[0,1,2]]]
A,B,C=[acos((y*y+z*z-x*x)/2/z/y)for(x,y,z)in[(a,b,c),(b,c,a),(c,a,b)]]
R=a/2/sin(A)
def g(x,y):return x if y<1e-3 else g(y,fmod(x,y))
u=2*g(A,g(B,C))
print(round(R*R*pi/u*sin(u),7))

File: direct/1C.clj
(ns polygon-area
  (:import [java.util Scanner]))

(def eps 1e-3)

;; floating point "gcd" using fmod
(defn fmod [x y]
  (- x (* y (Math/floor (/ x y)))))

(defn gcd [x y]
  (if (< y eps)
    x
    (recur y (fmod x y))))

(defn hypot [x y]
  (Math/hypot x y))

(defn -main []
  (let [scan (Scanner. System/in)
        ;; read three points
        p (repeatedly 3 #(vector (.nextDouble scan) (.nextDouble scan)))
        ;; compute side lengths
        sides (map (fn [[p1 p2]]
                     (apply hypot (map - p1 p2)))
                   [[(nth p 0) (nth p 1)]
                    [(nth p 1) (nth p 2)]
                    [(nth p 2) (nth p 0)]])
        [a b c] sides
        ;; compute angles
        angles (map (fn [[x y z]]
                      (Math/acos (/ (+ (* y y) (* z z) (- (* x x))) (* 2 y z))))
                    [[a b c] [b c a] [c a b]])
        [A B C] angles
        R (/ a (* 2 (Math/sin A)))
        u (* 2 (gcd A (gcd B C)))
        area (* R R Math/PI (/ (Math/sin u) u))]
    (println (format "%.7f" area))))

;; call main
(-main)
